### ISSUE TITLE
GCC jitter imul instruction

### ISSUE THREAD

**Issue by mrphrazer** [cea-sec/miasm#418]
Hi,

it seems that the gcc jitter has problems with `imul` and 64 bit registers, e.g., 

```asm
IMUL RSI, R8
```

This is what happens:

```python
/tmp/tmpomTLQI.c: In function ‘block_loc_0000000000626994’:
/tmp/tmpomTLQI.c:839:26: error: ‘uint128_t’ undeclared (first use in this function)
  var_08_00 = ((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)&0xffffffffffffffffffffffffffffffff)+(-(((((uint128_t)((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                          ^~~~~~~~~
/tmp/tmpomTLQI.c:839:26: note: each undeclared identifier is reported only once for each function it appears in
/tmp/tmpomTLQI.c:839:174: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)&0xffffffffffffffffffffffffffffffff)+(-(((((uint128_t)((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:839:363: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)&0xffffffffffffffffffffffffffffffff)+(-(((((uint128_t)((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:839:400: warning: integer constant is too large for its type
 & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)&0xffffffffffffffffffffffffffffffff)+(-(((((uint128_t)((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:839:436: warning: integer constant is too large for its type
 fffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)&0xffffffffffffffffffffffffffffffff)+(-(((((uint128_t)((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:839:647: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:839:836: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:839:873: warning: integer constant is too large for its type
 & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:839:1140: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:839:1329: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:839:1366: warning: integer constant is too large for its type
 & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:839:1496: warning: integer constant is too large for its type
 >63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:839:1533: warning: integer constant is too large for its type
  0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:840:174: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)&0xffffffffffffffffffffffffffffffff)+(-(((((uint128_t)((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:840:363: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)&0xffffffffffffffffffffffffffffffff)+(-(((((uint128_t)((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:840:400: warning: integer constant is too large for its type
 & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)&0xffffffffffffffffffffffffffffffff)+(-(((((uint128_t)((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:840:436: warning: integer constant is too large for its type
 fffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)&0xffffffffffffffffffffffffffffffff)+(-(((((uint128_t)((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:840:647: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:840:836: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:840:873: warning: integer constant is too large for its type
 & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)(((((((((((((uint128_t)(mycpu->RDX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RDX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:840:1140: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:840:1329: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:840:1366: warning: integer constant is too large for its type
 & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF)>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:840:1496: warning: integer constant is too large for its type
 >63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:840:1533: warning: integer constant is too large for its type
  0xFFFFFFFFFFFFFFFF)) << 64)))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)?0x1:0x0))&0x1;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:841:172: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff)*(((((uint128_t)(mycpu->RBX & 0xFFFFFFFFFFFFFFFF)) << 0) | (((uint128_t)((((mycpu->RBX>>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF))&0xFFFFFFFFFFFFFFFF;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:841:361: warning: integer constant is too large for its type
 >>63) & 0x1)?0xffffffffffffffff:0x0) & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF))&0xFFFFFFFFFFFFFFFF;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/tmpomTLQI.c:841:398: warning: integer constant is too large for its type
 & 0xFFFFFFFFFFFFFFFF)) << 64))&0xffffffffffffffffffffffffffffffff))&0xffffffffffffffffffffffffffffffff)>>0) & 0xFFFFFFFFFFFFFFFF))&0xFFFFFFFFFFFFFFFF;
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Traceback (most recent call last):
  File "emulate.py", line 111, in <module>
    sb.run()
  File "/usr/lib/python2.7/site-packages/miasm2/analysis/sandbox.py", line 481, in run
    super(Sandbox_Linux_x86_64, self).run(addr)
  File "/usr/lib/python2.7/site-packages/miasm2/analysis/sandbox.py", line 127, in run
    self.jitter.continue_run()
  File "/usr/lib/python2.7/site-packages/miasm2/jitter/jitload.py", line 363, in continue_run
    return self.run_iterator.next()
  File "/usr/lib/python2.7/site-packages/miasm2/jitter/jitload.py", line 338, in runiter_once
    self.pc = self.runbloc(self.pc)
  File "/usr/lib/python2.7/site-packages/miasm2/jitter/jitload.py", line 299, in runbloc
    return self.jit.runbloc(self.cpu, self.vm, pc, self.breakpoints_handler.callbacks)
  File "/usr/lib/python2.7/site-packages/miasm2/jitter/jitcore.py", line 187, in runbloc
    self.disbloc(lbl, vm)
  File "/usr/lib/python2.7/site-packages/miasm2/jitter/jitcore.py", line 162, in disbloc
    self.add_bloc(cur_bloc)
  File "/usr/lib/python2.7/site-packages/miasm2/jitter/jitcore_gcc.py", line 163, in add_bloc
    check_call(args)
  File "/usr/lib/python2.7/subprocess.py", line 541, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['gcc', '-O3', '-shared', '-fPIC', '/tmp/tmpomTLQI.c', '-o', '/tmp/tmpsQ0v5u.so', '-I/usr/lib/python2.7/site-packages/miasm2/jitter', '-I/usr/include/python2.7', '/usr/lib/python2.7/site-packages/miasm2/jitter/VmMngr.so', '/usr/lib/python2.7/site-packages/miasm2/jitter/arch/JitCore_x86.so']' returned non-zero exit status 1
```

**Comment by serpilliere** [comment]
You are right: but this was already the case using `TCC`. When the multiplication is done, internally it does the 64*64 =>128 bit, so it split the result as low and high part, to set the destination registers.


The trick is that in C, we don't have an `uint128_t`.
Anyway, it seems there is  a trick on `GCC` to create such an alias, so this may be added in the `GCC` back and. (but sadly the trick doesn't work under `tcc`)

Here it is:
```C
typedef unsigned uint128_t __attribute__ ((mode (TI)))
```

**Comment by mrphrazer** [comment]
Where do you think should that be placed? In `Jitgcc.c` it does not have any effect.

**Comment by mrphrazer** [comment]
oh, it appears that `vm_mngr.h` does the job. great! :)

**Comment by commial** [comment]
It seems to be a duplicate of #28 , so I'm closing this issue. Don't hesitate to re-open it if I'm wrong.
